dacite/types.py

```diff
@@ -86,6 +87,31 @@ def is_init_var(type_: Type) -> bool:
     return isinstance(type_, InitVar) or type_ is InitVar
 
 
+@cache
+def is_generic_alias(type_: Type) -> bool:
+    """Since `typing._GenericAlias` is not explicitly exported, we instead rely on this check."""
+    return str(type_) == "<class 'typing._GenericAlias'>"
+
+
+@cache
+def has_generic_alias_in_args(type_: Type) -> bool:
+    return is_generic_alias(type(get_args(type_)))
+
+
+def is_valid_generic_class(value: Any, type_: Type) -> bool:
+    origin = get_origin(type_)
+    if not (origin and isinstance(value, origin)):
+        return False
+    type_hints = cache(get_type_hints)(type(value))
+    for field_name, field_type in type_hints.items():
+        if isinstance(field_type, TypeVar):
+            args = get_args(type_)
+            return True if not args else any(isinstance(getattr(value, field_name, None), arg) for arg in args)
+        else:
+            return isinstance(value, type_)
```

Comment:
I'm using this branch for a project and found this. This block doesn't make sense and also doesn't work.

`get_type_hints` gets a mapping from field name to field type for the dataclass `value`. You're iterating over that, but you only ever check the first item because both branches end with `return`. I'm pretty sure both of those returns in the loop should only be returning if the value is False.

On the first branch, if `field_type` is a TypeVar, you check the field value against all the arguments for the subscripted generic `type_`. Why is `get_args` in the loop? Why is `getattr` in the inner loop? Neither one is changing at that point. I believe this would also fail to detect an incorrect type if, say, a field is hinted `AnyStr`, the value is None, and the dataclass has two type arguments, one `str` and one `None`.

I assume the second branch is trying to check if the *field* value matches the *field* type? But what you're actually doing is checking if the *dataclass* matches the *subscripted dataclass type*. This crashes with the exception `TypeError: Subscripted generics cannot be used with class and instance checks` for obvious reasons.

For the false negative issue with multiple type arguments, possibly something could be done with `__orig_bases__`? Not sure.

Here's my suggestion for a fixed version. I haven't written test cases, but I did try it with my fairly complicated use case and it seems to work.
```suggestion
    type_args = get_args(type_)
    type_hints = cache(get_type_hints)(type(value))
    for field_name, field_type in type_hints.items():
        field_value = getattr(value, field_name, None)
        if isinstance(field_type, TypeVar):
            # TODO: this will fail to detect incorrect type in some cases
            # see comments on https://github.com/konradhalas/dacite/pull/209
            if not any(is_instance(field_value, arg) for arg in type_args):
                return False
        elif get_origin(field_type) is not ClassVar:
            if not is_instance(field_value, field_type):
                return False
```